File: backend/routes/response_models/alert.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class AlertRuleCreate(BaseModel):
    sensor_id: int
    name: str
    description: Optional[str] = None
    condition: str
    threshold_lo: Optional[float] = None
    threshold_hi: Optional[float] = None
    recovery_lo: Optional[float] = None
    recovery_hi: Optional[float] = None
    severity: str = "warning"
    dwell_seconds: int = 60
    no_data_seconds: Optional[int] = None
    recovery_dwell_seconds: int = 30
    policy: str = "auto_clear"
    is_enabled: bool = True


class AlertRuleUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    condition: Optional[str] = None
    threshold_lo: Optional[float] = None
    threshold_hi: Optional[float] = None
    recovery_lo: Optional[float] = None
    recovery_hi: Optional[float] = None
    severity: Optional[str] = None
    dwell_seconds: Optional[int] = None
    no_data_seconds: Optional[int] = None
    recovery_dwell_seconds: Optional[int] = None
    policy: Optional[str] = None
    is_enabled: Optional[bool] = None


class AlertRuleRead(BaseModel):
    id: int
    sensor_id: int
    name: str
    description: Optional[str] = None
    condition: str
    threshold_lo: Optional[float] = None
    threshold_hi: Optional[float] = None
    recovery_lo: Optional[float] = None
    recovery_hi: Optional[float] = None
    severity: str
    dwell_seconds: int
    no_data_seconds: Optional[int] = None
    recovery_dwell_seconds: int
    policy: str
    is_enabled: bool
    created_by: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class AlertEventRead(BaseModel):
    id: int
    rule_id: int
    sensor_id: int
    state: str
    observed_value: Optional[float] = None
    observed_at: datetime
    acknowledged_by: Optional[int] = None
    acknowledged_at: Optional[datetime] = None
    note: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/app/types/alert.ts
export interface AlertRule {
  id: number;
  sensor_id: number;
  name: string;
  description: string | null;
  condition: 'gt' | 'lt' | 'outside_range' | 'inside_range' | 'no_data';
  threshold_lo: number | null;
  threshold_hi: number | null;
  recovery_lo: number | null;
  recovery_hi: number | null;
  severity: 'warning' | 'critical';
  dwell_seconds: number;
  no_data_seconds: number | null;
  recovery_dwell_seconds: number;
  policy: 'auto_clear' | 'manual_ack';
  is_enabled: boolean;
  created_by: number | null;
  created_at: string;
  updated_at: string;
}

export interface AlertEvent {
  id: number;
  rule_id: number;
  sensor_id: number;
  state:
    | 'pending'
    | 'firing'
    | 'recovered'
    | 'awaiting_ack'
    | 'cleared'
    | 'flapping_started'
    | 'flapping_stopped';
  observed_value: number | null;
  observed_at: string;
  acknowledged_by: number | null;
  acknowledged_at: string | null;
  note: string | null;
  created_at: string;
}

export interface AlertRoute {
  id: number;
  scope_kind: 'rule' | 'sensor' | 'asset' | 'severity' | 'all';
  scope_id: number | null;
  scope_severity: string | null;
  channel: 'in_app' | 'email' | 'webhook';
  target: string;
  on_fire: boolean;
  on_recover: boolean;
  is_enabled: boolean;
  created_at: string;
}
